### PR TITLE
[sk] Added frozen dictionary type to handle lists with dictionary objects

### PR DESCRIPTION
**Pull request by skunichetty**
# Summary
A small edge case occurred when a list of dictionary objects was stored within a Pandas data frame. While support for lists was handled before, handling dictionary objects within these lists causes various problems throughout the data cleaning pipeline due to an inability to hash a dictionary (including in statistics calculation and graph generation). 

To solve this problem a `FrozenDict` type was created, an immutable key-value store that doesn't permit mutations. For this reason, this type can be hashable without any major concerns of mutability. All non-mutating operations are still supported (such as fetching values from the dictionary).

When parsing a dictionary item in a list, the dictionary is converted to a `FrozenDict`, allowing the dictionary objects to be handled by Pandas without any errors.

# Tests
Changes were tested locally alongside writing new unit test cases.

cc:
@wangxiaoyou1993

### CODE REVIEW

File: mage_ai/data_cleaner/shared/utils.py [hunk 1]
```diff
@@ -1,6 +1,6 @@
 from mage_ai.data_cleaner.column_types.constants import NUMBER_TYPES, ColumnType
+from mage_ai.shared.custom_types import FrozenDict
```
[the remark this answers is not in the document]
Comment (skunichetty) [reply mage_ai/data_cleaner/shared/utils.py]: Fixed.

File: mage_ai/data_cleaner/shared/utils.py [hunk 2]
```diff
@@ -88,16 +88,25 @@ def __parse_element(element: str) -> Any:
         return np.nan
     else:
         try:
-            return eval(element)
+            parsed_element = eval(element)
```
[the remark this answers is not in the document]
Comment (skunichetty) [reply mage_ai/data_cleaner/shared/utils.py]: Missed that optimization, it's fixed now.

File: mage_ai/data_cleaner/shared/utils.py [hunk 3]
```diff
@@ -88,16 +88,25 @@ def __parse_element(element: str) -> Any:
...
 
 
 def parse_list(list_literal: Union[str, List[Any]]) -> FrozenList:
     dtype = type(list_literal)
     if dtype is FrozenList:
         return list_literal
     elif dtype in LIST_TYPES:
+        if any(not pd.api.types.is_hashable(literal) for literal in list_literal):
+            return FrozenList([__resolve_type(literal) for literal in list_literal])
         return FrozenList(list_literal)
```
[the remark this answers is not in the document]
Comment (skunichetty) [reply mage_ai/data_cleaner/shared/utils.py]: Agreed, I thought I could reduce the amount of work done here but it seems like regardless the entire list will be traversed with the same type of filtering. I changed this function so the `__resolve_type` logic is always applied.

File: mage_ai/shared/custom_types.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+            return self[__k]
+        else:
+            raise AttributeError('Cannot assign a value to frozen dictionary')
+
+    def update(self, other=None) -> Any:
+        raise AttributeError('Cannot assign a value to frozen dictionary')
+
+    def __hash__(self):
+        # TODO: Come up with better hash function
+        return hash(tuple((key, str(value)) for key, value in self.items()))
```
Comment (dy46) [inline mage_ai/shared/custom_types.py:31]: is there a high chance of any overlap or issues like that?

**Review by dy46 — APPROVED** [review]

File: mage_ai/shared/custom_types.py (hunk 4 above)
Comment (skunichetty) [reply mage_ai/shared/custom_types.py:31]: From what I can tell, there shouldn't be a high chance of overlap or collision. Hashing tuples preserves order, so the order of the entries in the dictionary will not cause collisions often. So hashing a tuple of `(key, value)` tuples should generate a fairly unique hash for the object. I'm just unsure if there is a better method.

**Pull request merged**